### PR TITLE
Converted Simulated HRVO tests to use PyTest

### LINKED ISSUE #2759: Port Simulated HRVO tests to use the new PyTest testing framework
### Description of the task
Currently our simulated HRVO tests are all written using our old C++ testing framework which is much more limited than our new PyTest testing framework (e.g. Can't replay tests, validations are not visualized, ...). As a part of this task, the tests should be ported over and should be parameterized as much as possible to avoid lots of duplications and long/unreadable test files.
For some examples of simulated tests written using PyTest, have a look at `goalie_tactic_test.py` or #2754. I have also played around with some simulated HRVO tests in Python in a test branch here (note that this is not complete and should be extended to allow for more robots and more complex scenarios): https://github.com/nimazareian/Software/blob/nima/ramping_test_in_sim2/src/extlibs/hrvo/hrvo_test.py

### Acceptance criteria
- [ ] Port over simulated HRVO tests to the new testing framework, removing all test cases which seem unnecessary.
- [ ] Add tests for any scenario not tested (e.g. We currently don't have a test where we try to avoid obstacles other than robots. We should add tests where we avoid the ball, either as a half meter obstacle or just as it self (i.e. `ball_collision_type=BallCollisionType.AVOID`))
- [ ] Parameterize the tests to reduce the amount of duplicated code
- [ ] Validate that robots are moving to their desired destination. Some other possible validations could include checking for illegal collisions (SSL rules specify and minimum speed for a collision to be a foul).

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

Simulated HRVO tests were using the .cpp test framework before. Converted them to use the new PyTest simulated test fixture

### Testing Done

All tests pass and the expected behaviour is displayed on Thunderscope

### Resolved Issues

Resolves #2759

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by sauravbanna** [comment]
Mainly focuses on simulated_hrvo_test.py

File: src/software/ai/navigator/path_planner/hrvo/BUILD [hunk 1]
```diff
@@ -64,7 +66,7 @@ cc_test(
 )
 
 cc_test(
-    name = "simulated_hrvo_test",
+    name = "simulated_hrvo_test_old",
     srcs = ["simulated_hrvo_test.cpp"],
```
Comment (nimazareian) [inline src/software/ai/navigator/path_planner/hrvo/BUILD]: Since youve ported over the tests, you can remove the old C++ tests!

File: src/software/ai/navigator/path_planner/hrvo/hrvo_agent.cpp [hunk 2]
```diff
@@ -225,6 +225,36 @@ VelocityObstacle HRVOAgent::createVelocityObstacle(const Agent &other_agent)
     }
 
     return VelocityObstacle(hrvo_apex, vo.getLeftSide(), vo.getRightSide());
+
+    //    const double combinedRadius = radius * 2;
```
Comment (nimazareian) [inline src/software/ai/navigator/path_planner/hrvo/hrvo_agent.cpp]: commented out code

File: src/software/ai/hl/stp/tactic/dribble/dribble_tactic_test.cpp [hunk 3]
```diff
@@ -52,75 +52,75 @@ class DribbleTacticTest : public SimulatedErForceSimPlayTestFixture
...
-        }};
-
-    std::vector<ValidationFunction> non_terminating_validation_functions = {};
-
-    runTest(field_type, ball_state, friendly_robots, enemy_robots,
-            terminating_validation_functions, non_terminating_validation_functions,
-            Duration::fromSeconds(10));
-}
-
+//TEST_F(DribbleTacticTest, test_intercept_ball_behind_enemy_robot)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/tactic/dribble/dribble_tactic_test.cpp]: Are these tests supposed to be commented out?

File: src/software/simulated_tests/avoid_collisions.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import software.python_bindings as tbots
+from proto.import_all_protos import *
+from software.simulated_tests.validation import (
+    Validation,
+    create_validation_types,
+    create_validation_geometry,
+)
+
+
+class RobotDoesNotCollide(Validation):
```
Comment (nimazareian) [inline src/software/simulated_tests/avoid_collisions.py]: Nit: This validation name may be slightly confusing. Could be more descriptive if the name is plural (i.e. robotics not colliding with each other, rather than colliding with some general obstacle)
```suggestion
class RobotsDoNotCollide(Validation):
```

File: src/software/simulated_tests/avoid_collisions.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        robot_2_vel = tbots.createVector(
+            Vector(
+                x_component_meters=robot2.current_state.global_velocity.x_component_meters,
+                y_component_meters=robot2.current_state.global_velocity.y_component_meters,
+            )
+        )
+
+        # check if robots are colliding
+        if (robot_1_pos - robot_2_pos).length() < ROBOT_MAX_RADIUS_METERS * 2 + 0.005:
```
Comment (nimazareian) [inline src/software/simulated_tests/avoid_collisions.py]: nit: lets refactor the `0.005` into a constant variable

File: src/software/simulated_tests/avoid_collisions.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            # check if the projection of the difference between their velocities
+            # onto the line between their positions > 1.5
+            # if this case is true, robots are considered to have collided
+            # but still have to determine which ones have fouled
+            if (robot_1_vel - robot_2_vel).dot(robot_1_pos - robot_2_pos) / (
+                robot_1_pos - robot_2_pos
+            ).length() > 1.5:
+                # checks if the absolute speed difference between the robots is < 0.3
+                # if so, both get a foul
+                if robot_1_vel.length() - robot_2_vel.length() < 0.3:
```
Comment (nimazareian) [inline src/software/simulated_tests/avoid_collisions.py]: We should be taking the absolute difference
```suggestion
                if abs(robot_1_vel.length() - robot_2_vel.length()) < 0.3:
```

File: src/software/simulated_tests/avoid_collisions.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                    self.fouled_robots.extend([robot1.id, robot2.id])
+                else:
+                    # checks which robot is faster (that one will get a foul)
+                    if robot_1_vel.length() > robot_2_vel.length():
+                        self.fouled_robots.extend([robot1.id])
+                    else:
+                        self.fouled_robots.extend([robot2.id])
+                return True
```
Comment (nimazareian) [inline src/software/simulated_tests/avoid_collisions.py]: Could you also link the SSL rule where this logic is from: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_crashing

File: src/software/simulated_tests/robot_enters_region_and_stops.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            if robot.id == self.robot_id:
+                if tbots.contains(
+                    self.region, tbots.createPoint(robot.current_state.global_position)
+                ):
+                    if (
+                        math.hypot(
+                            robot.current_state.global_velocity.x_component_meters,
+                            robot.current_state.global_velocity.y_component_meters,
+                        )
+                        < 0.01
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_region_and_stops.py]: Can we refactor this into a constant with units in the name

File: src/software/simulated_tests/robot_enters_region_and_stops.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        Returns a string representing the stage of validation that failed
+        Either the robot has not entered the region yet, or it has but is not stationary
+        """
+        if not self.is_stationary:
+            return f"Check that robot {self.robot_id} in correct region is stationary"
+
+        return (
+            f"Check for stationary robot {self.robot_id} in region "
+            + ","
+            + repr(self.region)
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_region_and_stops.py]: I'm not exactly sure how we are generally trying to have these string messages, but for this case, could you also add that the robot is stationary, but not contained inside the region? It'd be nice to have these include as much detail as possible for when they fail

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+            ],
+            [
+                pos
+                for index, pos in enumerate(get_robot_circle_pos(1.5, 8, False))
+                if index % 2 == 0
+            ],
+            [],
+            # both the start and end positions are the same values
+            # but for some reason it works when actually running the test
+            # so just gonna leave this here
```
Comment (nimazareian) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py]: It would be good to figure this out if it's not too complex just to clear any future confusions with the test setups

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+def get_robot_circle_pos(radius, num_robots, start):
+    """
+    Gets the test params to position robots in a circle and have them move
+    along each diameter
+
+    :param radius: the radius of the circle
+    :param num_robots: the number of robots in the circle
+    :param start: if True, start positioning the robots from (radius, 0)
+                  else, start from (-radius, 0)
+                  basically outputs robot start and end points based on boolean
```
Comment (nimazareian) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py]: Curious why this is necessary?

**Review by nimazareian — COMMENTED** [review]
I know that this PR is still a draft, but I had some time so I thought giving early feedback may be helpful.

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py (hunk 11 above)
Comment (sauravbanna) [reply src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py]: so this basically means that calling the function with start == True gives you the start positions of the robots. And calling it with start == False gives you the end positions of the same robots, which are directly opposite each start position.

this is basically so that the same function can be used for both. the boolean rotates the circle of positions generated by 180 degrees

File: src/software/simulated_tests/robot_enters_region_and_stops.py (hunk 9 above)
Comment (sauravbanna) [reply src/software/simulated_tests/robot_enters_region_and_stops.py]: so the thing is the validation always checks for being in the correct region first, then worries about if its stationary or not. So a test would never fail with the robot being stationary while not being in the correct region because the stationary check happens after the region check.

File: src/software/ai/navigator/path_planner/hrvo/BUILD (hunk 1 above)
Comment (sauravbanna) [reply src/software/ai/navigator/path_planner/hrvo/BUILD]: oh yeah sure, I'll remove them once the pytest validation is finalised

File: src/software/simulated_tests/avoid_collisions.py (hunk 5 above)
Comment (nimazareian) [reply src/software/simulated_tests/avoid_collisions.py]: The constant can be defined in the same file

File: src/shared/constants.h [hunk 12]
```diff
@@ -165,3 +165,6 @@ static const unsigned NUM_TICKS_PER_TRAJECTORY_ELEMENT = 4u;
 static const unsigned CONTROL_LOOP_HZ                  = 60u;
 
 static const unsigned NUM_GENEVA_ANGLES = 5;
+
+static const double ROBOT_COLLISION_BUFFER             = 0.005;
+static const double ROBOT_MAX_STATIONARY_SPEED_M_PER_S = 0.01;
```
Comment (nimazareian) [inline src/shared/constants.h]: Constants that are specific to Python should just be define in Python. This way also if we change them, then a bunch of C++ code doesn't have to recompile

**Review by nimazareian — COMMENTED** [review]
overall looks good, just a nit

File: src/shared/constants.h (hunk 12 above)
Comment (itsarune) [reply src/shared/constants.h]: Ideally we wouldn't mix test and robot constants together, so I agree with nima

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -103,7 +105,8 @@ jobs:
           bazel test --show_timestamps       \
             //software:unix_full_system           \
             //software/simulated_tests/...   \
-            //software/ai/hl/...
+            //software/ai/hl/... \
+            //software/ai/navigator/...
```
Comment (itsarune) [inline .github/workflows/main.yml:135]: you need to make this change one more time for "Simulated Tests 3"

Comment (sauravbanna) [reply .github/workflows/main.yml:135]: done

File: src/software/simulated_tests/avoid_collisions.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+                # if so, both get a foul
+                if abs(robot_1_vel.length() - robot_2_vel.length()) < 0.3:
+                    self.fouled_robots.extend([robot1.id, robot2.id])
+                else:
+                    # checks which robot is faster (that one will get a foul)
+                    if robot_1_vel.length() > robot_2_vel.length():
+                        self.fouled_robots.extend([robot1.id])
+                    else:
+                        self.fouled_robots.extend([robot2.id])
+                return True
```
Comment (Bvasilchikov) [inline src/software/simulated_tests/avoid_collisions.py]: Separate this out into a another function call. The function signature doesn't reflect that it checks for fouls. Generally you should avoid putting side-effecting code `self.fouled_robots.extend(...)` and have functions focus on a single task. This allows the caller to have better control and reduces bugs.

**Review by Bvasilchikov — DISMISSED** [review]
lgtm! just 1 comment

File: src/software/simulated_tests/avoid_collisions.py (hunk 14 above)
Comment (itsarune) [inline src/software/simulated_tests/avoid_collisions.py]: this is awesome validation to have!

File: src/software/simulated_tests/robot_enters_region_and_stops.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        :param robot_id: the id of the robot being validated
+        :param region: the region the robot has to enter
+        :param num_ticks: the time the robot has to stay stationary for in the regions
+        """
+        self.region = region
+        self.robot_id = robot_id
+        self.num_ticks = num_ticks
+        self.ticks_so_far = 0
+
+        self.is_stationary = True
```
Comment (itsarune) [inline src/software/simulated_tests/robot_enters_region_and_stops.py]: One of the goals of the new validation function system is to be robot id agnostic so that we don't depend on changes to the robot to tactic assignment algorithm.

Instead of this new validation function, I recommend extending the RobotEntersRegion validation function to support checking if the robot is stopped. Check if any robot is in the region and return PASSING instead of just a specific robot.

File: src/software/simulated_tests/avoid_collisions.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+    create_validation_geometry,
+)
+
+
+class RobotsDoNotCollide(Validation):
+    """
+    Checks if any 2 robots have collided
+    """
+
+    ROBOT_COLLISION_BUFFER = 0.005
```
Comment (itsarune) [inline src/software/simulated_tests/avoid_collisions.py]:
```suggestion
    ROBOT_COLLISION_BUFFER_M = 0.005
```

**Review by itsarune — COMMENTED** [review]
These are really awesome tests!

Since you wrote these tests to be very portable, we should be able to use these tests very easily for the new motion planner so let's try to merge it!

File: src/software/simulated_tests/avoid_collisions.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+        return f"Check that no robots are colliding with each other"
+
+
+(
+    RobotEventuallyDoesNotCollide,
+    RobotEventuallyCollides,
+    RobotAlwaysDoesNotCollide,
+    RobotAlwaysCollides,
+) = create_validation_types(RobotsDoNotCollide)
```
Comment (nimazareian) [inline src/software/simulated_tests/avoid_collisions.py]: I think the generated validation types should be plural

File: src/software/simulated_tests/tbots_test_runner.py [hunk 18]
```diff
@@ -140,6 +140,15 @@ def set_play(self, play: Play, is_friendly: bool):
...
         fs_proto_unix_io.send_proto(Play, play)
 
+    def clear_world_buffer(self):
+        """
+        Clears the world buffer
+        """
+        world = self.world_buffer.get(return_cached=False)
+
+        while world is not None:
+            world = self.world_buffer.get(return_cached=False)
```
Comment (nimazareian) [inline src/software/simulated_tests/tbots_test_runner.py]: Why is this needed? I noticed it being used in ball placement play test, though that test should have been fixed after your previous PR.

File: src/software/simulated_tests/terminating_validation_functions/robot_stationary_in_polygon_validation.cpp
Comment (nimazareian) [inline src/software/simulated_tests/terminating_validation_functions/robot_stationary_in_polygon_validation.cpp:1]: Since we have a python equivalent to this and this validation is not used anymore + we shouldn't be adding any more C++ simulated tests, could you remove this file (including from the BUILD file)

**Review by nimazareian — COMMENTED** [review]
Just a couple of final nits and questions. Also, it seems like you forgot to push your fix to this error msg:
```
...
gl_validation_layer.py", line 119, in <listcomp>
    for polygon in validation.geometry.polygons
AttributeError: 'tuple' object has no attribute 'geometry'
```

File: src/software/simulated_tests/robot_enters_region.py [hunk 19]
```diff
@@ -27,8 +28,10 @@ def get_validation_status(self, world) -> ValidationStatus:
                 if tbots_cpp.contains(
                     region, tbots_cpp.createPoint(robot.current_state.global_position)
                 ):
+                    self.passing_robot = robot
```
Comment (itsarune) [inline src/software/simulated_tests/robot_enters_region.py:31]: it doesn't look like self.passing_robot is being used

Comment (itsarune) [reply src/software/simulated_tests/robot_enters_region.py:31]: oh it's used in the subclass

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+        # walls generated by range(start_x, start_x + step * num_robots, step)
+        (
+            [tbots.Point(2.8, -1)],
+            [tbots.Point(2.8, 1)],
+            [],
+            [tbots.Point(float(x_meters) / 10, 0) for x_meters in range(20, 32, 2)],
+            [],
+            25,
+            True,
+        ),
```
Comment (itsarune) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:349]: the test where there is a wall of enemy robots in the enemy half and the robot has to travel across doesn't reach the destination and still passes. Is this intentional?

**Review by itsarune — COMMENTED** [review]
we should expect that a robot that doesn't reach the destination to fail the test right?

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py (hunk 20 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:349]: For some of the tests there's no validation for reaching the destination since they're more just for observation of what happens on edge cases and are not necessirally deterministic well.

**Review by nimazareian — APPROVED** [review]

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 21]
```diff
@@ -0,0 +1,566 @@
...
+            + repr(self.region)
+        )
+
+
+(
+    HRVORobotEventuallyEntersRegionAndStops,
+    HRVORobotEventuallyExitsRegionAndMoves,
+    HRVORobotAlwaysStaysInRegionStationary,
+    HRVORobotNeverEntersRegionAndMoves,
+) = create_validation_types(HRVORobotEntersRegionAndStops)
```
Comment (Bvasilchikov) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:111]: Can you move these classes outside?

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 22]
```diff
@@ -0,0 +1,566 @@
...
+
+
+def get_zig_zag_params(
+    front_wall_x: int, robot_y_delta: float, num_walls: int, wall_height: float
+):
+    """
+    Gets the test params to cause movement in a zig zag pattern
+    due to multiple walls of enemy robots
```
Comment (Bvasilchikov) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:119]:
```suggestion
def create_zig_zag_path_test(
    front_wall_x: int, robot_y_delta: float, num_walls: int, wall_height: float
):
    """
    Creates a test that positions enemy robots in rows of walls, as well as start and destination points
    that encourage a friendly robot to zig-zag through the walls as the most optimal path.
```

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py [hunk 23]
```diff
@@ -0,0 +1,566 @@
...
+        [],
+        20,
+        False,
+    )
+
+
+def get_robot_circle_pos(radius: float, num_robots: int, start: bool):
+    """
+    Gets the test params to position robots in a circle and have them move
+    along each diameter
```
Comment (Bvasilchikov) [inline src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:152]:
```suggestion
def create_robot_circle_pos(radius: float, num_robots: int, start: bool):
    """
    Constructs a list of points that evenly positions num_robots points
    along the diameter of a circle radius r
```

**Review by JoshuaWu7 — COMMENTED** [review]
looks good to me, ~just pending @itsarune's comment above~ I see Nima has already answered the question above thanks!

**Review by JoshuaWu7 — APPROVED** [review]

**Review by Bvasilchikov — APPROVED** [review]
Great work converting all the tests and adding support for new ones!

Had a few nits if you wanna take a look, but lets get this merged in.

File: src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py (hunk 21 above)
Comment (sauravbanna) [reply src/software/ai/navigator/path_planner/hrvo/simulated_hrvo_test.py:111]: they were outside lol, but arun said our validations shouldn't be robot id dependant. but since this one has to be, I moved it inside and it's only gonna be used for these tests
